nats/js/client.py

```diff
@@ -136,6 +185,50 @@ async def publish(
             raise nats.js.errors.APIError.from_error(resp['error'])
         return api.PubAck.from_response(resp)
 
+    async def publish_async(
+        self,
+        subject: str,
+        payload: bytes = b'',
+        stream: Optional[str] = None,
+        headers: Optional[Dict] = None,
+    ) -> asyncio.Future[api.PubAck]:
+        """
+        emits a new message to JetStream and returns a future that can be awaited for acknowledgement.
+        """
+
+        if len(self._paf_map) > self._pending_acks_limit:
+            raise nats.js.errors.TooManyStalledMsgsError
+
+        if not self._paf_sub_prefix:
+            await self._init_paf_sub()
+        assert self._paf_sub_prefix
+
+        hdr = headers
+        if stream is not None:
+            hdr = hdr or {}
+            hdr[api.Header.EXPECTED_STREAM] = stream
+
+        # Use a new NUID + couple of unique token bytes to identify the request,
+        # then use the future to get the response.
+        token = self._nc._nuid.next()
+        token.extend(token_hex(2).encode())
+        inbox = self._paf_sub_prefix[:]
+        inbox.extend(token)
+        future: asyncio.Future = asyncio.Future()
+        self._paf_map[token.decode()] = future
+
+        await self._nc.publish(
+            subject, payload, reply=inbox.decode(), headers=hdr
+        )
+
+        return future
+
+    async def publish_async_completed(self) -> None:
+        """
+        waits for all currently pending async publishes to be completed.
+        """
+        await asyncio.gather(*self._paf_map.values(), return_exceptions=True)
```

Comment:
might be good to also add `js.publish_async_pending` to get a snapshot of how many publish acks are inflight:
https://github.com/nats-io/nats.go/blob/main/js.go#L780-L784